ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -238,6 +292,190 @@ VALUE cpu_and_wall_time_collector_sample(VALUE self_instance) {
   return Qnil;
 }
 
+// This function gets called when Ruby is about to start running the Garbage Collector on the current thread.
+// It updates the per_thread_context of the current thread to include the current cpu/wall times, to be used to later
+// create a stack sample that blames the cpu/wall time spent from now until the end of the garbage collector work.
+//
+// Safety: This function gets called while Ruby is doing garbage collection. While Ruby is doing garbage collection,
+// *NO ALLOCATION* is allowed. This function, and any it calls must never trigger memory or object allocation.
+// This includes exceptions and use of ruby_xcalloc (because xcalloc can trigger GC)!
+//
+// Assumption 1: This function is called in a thread that is holding the Global VM Lock. Caller is responsible for enforcing this.
+void cpu_and_wall_time_collector_on_gc_start(VALUE self_instance) {
+  struct cpu_and_wall_time_collector_state *state;
+  if (!rb_typeddata_is_kind_of(self_instance, &cpu_and_wall_time_collector_typed_data)) return;
+  // This should never fail the the above check passes
+  TypedData_Get_Struct(self_instance, struct cpu_and_wall_time_collector_state, &cpu_and_wall_time_collector_typed_data, state);
+
+  struct per_thread_context *thread_context = get_context_for(rb_thread_current(), state);
+
+  // If there was no previously-existing context for this thread, we won't allocate one (see safety). For now we just drop
+  // the GC sample, under the assumption that "a thread that is so new that we never sampled it even once before it triggers
+  // GC" is a rare enough case that we can just ignore it.
+  // We can always improve this later if we find that this happens often (and we have the counter to help us figure that out)!
```

Comment:
This seems like a reasonable scenario what we can't record the GC sample.

Reply:
Indeed, that was my thinking!

Furthermore, in the future we'll want to integrate with the GVL tracing events, one of which is `RUBY_INTERNAL_THREAD_EVENT_STARTED` which is fired just after the thread is initialized.

Thus, if we do context initialization when that event gets triggered, this situation of "GC on a thread with no context" should be completely eliminated on Ruby 3.2+.